src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+            let second = &path[.. idx.unwrap_or(path.len())];
+            Some((first, second))
+        }
+    }
+
+    #[derive(Copy, Clone, PartialEq)]
+    pub enum Prefix<'a> {
+        Verbatim(&'a OsStr),
+        VerbatimUNC(&'a OsStr, &'a OsStr),
+        VerbatimDisk(&'a OsStr),
+        DeviceNS(&'a OsStr),
+        UNC(&'a OsStr, &'a OsStr),
+        Disk(&'a OsStr),
+    }
+
+    impl<'a> Prefix<'a> {
+        #[inline]
+        pub fn len(&self) -> usize {
+            use self::Prefix::*;
+            fn os_str_len(s: &OsStr) -> usize {
+                unsafe { os_str_as_u8_slice(s).len() }
+            }
+            match *self {
+                Verbatim(x) => 4 + os_str_len(x),
+                VerbatimUNC(x,y) => 8 + os_str_len(x) +
+                    if os_str_len(y) > 0 { 1 + os_str_len(y) }
+                    else { 0 },
+                VerbatimDisk(_) => 6,
+                UNC(x,y) => 2 + os_str_len(x) +
+                    if os_str_len(y) > 0 { 1 + os_str_len(y) }
+                    else { 0 },
+                DeviceNS(x) => 4 + os_str_len(x),
+                Disk(_) => 2
+            }
+
+        }
+
+        #[inline]
+        pub fn is_verbatim(&self) -> bool {
+            use self::Prefix::*;
+            match *self {
+                Verbatim(_) | VerbatimDisk(_) | VerbatimUNC(_, _) => true,
+                _ => false
+            }
+        }
+
+        #[inline]
+        pub fn is_drive(&self) -> bool {
+            match *self {
+                Prefix::Disk(_) => true,
+                _ => false,
+            }
+        }
+
+        #[inline]
+        pub fn has_implicit_root(&self) -> bool {
+            !self.is_drive()
+        }
+    }
+
+    pub const MAIN_SEP: u8 = b'\\';
+    pub const MAIN_SEP_STR: &'static str = "\\";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Misc helpers
+////////////////////////////////////////////////////////////////////////////////
+
+// Iterate through `iter` while it matches `prefix`; return `None` if `prefix`
+// is not a prefix of `iter`, otherwise return `Some(iter_after_prefix)` giving
+// `iter` after having exhausted `prefix`.
+fn iter_after<A, I, J>(mut iter: I, mut prefix: J) -> Option<I> where
+    I: Iterator<Item=A> + Clone, J: Iterator<Item=A>, A: PartialEq
+{
+    loop {
+        let mut iter_next = iter.clone();
+        match (iter_next.next(), prefix.next()) {
+            (Some(x), Some(y)) => {
+                if x != y { return None }
+            }
+            (Some(_), None) => return Some(iter),
+            (None, None) => return Some(iter),
+            (None, Some(_)) => return None,
+        }
+        iter = iter_next;
+    }
+}
+
+// See note at the top of this module to understand why these are used:
+unsafe fn os_str_as_u8_slice(s: &OsStr) -> &[u8] {
+    mem::transmute(s)
+}
+unsafe fn u8_slice_as_os_str(s: &[u8]) -> &OsStr {
+    mem::transmute(s)
+}
```

Comment:
Out of curiosity, how come these two functions are `unsafe`? It seems to require lots of unsafe blocks below, and they don't look necessarily inherently unsafe as they're just transmuting the representation to view it a different way (both read-only).